src/algebra/order/sub.lean

```diff
@@ -670,6 +670,21 @@ lemma tsub_right_inj (hba : b ≤ a) (hca : c ≤ a) : a - b = a - c ↔ b = c :
 contravariant.add_le_cancellable.tsub_right_inj contravariant.add_le_cancellable
   contravariant.add_le_cancellable hba hca
 
+variables (α)
+
+/-- A `canonically_ordered_add_monoid` with ordered subtraction and order-reflecting addition is
+cancellative. This is not an instance at it would form a typeclass loop.
+
+See note [reducible non-instances]. -/
+@[reducible]
+def canonically_ordered_add_monoid.to_add_cancel_comm_monoid : add_cancel_comm_monoid α :=
+{ add_left_cancel := λ a b c h, begin
+    have := congr_arg (λ x, x - a) h,
+    dsimp at this,
+    rwa [add_tsub_cancel_left, add_tsub_cancel_left] at this,
+  end,
```

Comment:
```suggestion
{ add_left_cancel := λ a b c h, by simpa only [add_tsub_cancel_left] using congr_arg (λ x, x - a) h,
```
untested

Reply:
I'm optimistic that will work, let's find out...